Move state-related utils to salt.utils.state.py

rallytime opened the pull request:
Moves the following functions from `salt.utils` to `salt.utils.state`:
- gen_state_tag
- search_onfail_requisites
- check_onfail_requisites
- check_state_result

Original functions have a deprecation warning added and all references in Salt code were updated to use the new function path, including moving some unit tests.

Commit:
Move state-related utils to salt.utils.state.py

Moves the following functions from `salt.utils` to `salt.utils.state`:
- gen_state_tag
- search_onfail_requisites
- check_onfail_requisites
- check_state_result

Original functions have a deprecation warning added and all references
in Salt code were updated to use the new function path, including moving
some unit tests.

salt-jenkins commented:
@rallytime, thanks for your PR! By analyzing the history of the files in this pull request, we identified @kiorky, @thatch45 and @terminalmage to be potential reviewers.

aneeshusa commented:
I'm definitely a +1 on having a `salt.utils.state` module (see #43096 for example). That being said, can you please use `__utils__` where possible to make it easier to backport and override modules?

rallytime commented:
@aneeshusa Yeah, for sure. Good idea!

Commit:
Change function names to not be so redundant

Now that the name of the util file is state.py, the function names can be
shortened:

- check_state_result --> check_result
- gen_state_tag --> gen_tag

Commit:
Use __utils__ where possible for moved state utils

And update all references to new state names in salt.utils.state.py

rallytime commented:
Ok, @aneeshusa I've updated the calls to use `__utils__` where possible.

I have also changed two of the functions names, at the recommendation of @terminalmage offline. Now that the functions are located in `salt.utils.state.py`, having a call like `state.check_state_result` is redundant. Two functions were renamed during this move:
- check_state_result --> check_result
- gen_state_tag --> gen_tag

aneeshusa commented on doc/ref/states/aggregate.rst:
```diff
@@ -122,7 +122,7 @@ This example, simplified from the pkg state, shows how to create mod_aggregate f
             # The state runtime uses "tags" to track completed jobs, it may
             # look familiar with the _|-
-            tag = salt.utils.gen_state_tag(chunk)
+            tag = salt.utils.state.gen_tag(chunk)
```
Since this would be in a state module, I think this can also use `__utils__`.

rallytime replied:
👍  Fixed

aneeshusa commented on salt/utils/state.py:
```diff
@@ -0,0 +1,155 @@
+def check_result(running, recurse=False, highstate=None):
+    '''
+    Check the total return value of the run and determine if the running
+    dict has any issues
```
This is a bit vague to me - any chance you can add a bit more description about what kinds of issues this checks for? (I'm curious how this differs from the `verify_ret_from_export` func in #43070, which I considered putting into `salt.utils.state`.)

rallytime replied:
@aneeshusa That's a good question that I'll let @thatch45 answer, since he was the original author of this function.

This function is used to check against the state result in many places across salt, and this PR is just moving the function to a new home as part of an on-going effort between @terminalmage and I to break up the massive `salt/utils/__init__.py` file in to smaller, more useful pieces.

aneeshusa replied:
Sure, I was just curious about this and probably best addressed in a follow-up, shouldn't block this PR going forward. Proceed with decomp!

rallytime replied:
Yeah, sounds good. Thank you for your review, as always. It's greatly appreciated. I'll try to swing by your other PR that also add the state util file soon. :)

aneeshusa reviewed (commented):
LGTM, had a couple of suggestions.

Commit:
Adjust test mocking to handle __utils__['state.check_result']

aneeshusa commented on salt/utils/__init__.py:
```diff
@@ -3403,3 +3255,89 @@ def get_colors(use=True, theme=None):
+    import salt.utils.state
+    salt.utils.versions.warn_until(
+        'Neon',
+        'Use of \'salt.utils.search_onfail_requisites\' detected. This function'
```
nit: In this and the next three calls to `warn_until`, add a trailing space like in `gen_state_tag` to prevent the words `function` and `has` from running together.

Commit:
Adjust a couple more tests to handle __utils__['state.check_result'] mocking

rallytime commented:
re-run py3

rallytime commented:
re-run py3
